### ISSUE TITLE
Vim9script: Syntax error goes unreported

### ISSUE THREAD

**Issue by girishji** [vim/vim#15936]
### Steps to reproduce

Run the following, and it reports no error:

```
❯ vim -Nu NONE -S <(cat <<EOF
    vim9script
    :set shortmess=I
    if true
        (false)  
        echom 'here'
    endif
    # vim: ft=vim sw=4 sts=4 et
EOF
)
```

Now, run this:

```
❯ vim -Nu NONE -S <(cat <<EOF
    vim9script
    :set shortmess=I
    if true
        false  
        echom 'here'
    endif
    # vim: ft=vim sw=4 sts=4 et
EOF
)
```
It will report `E492: Not an editor command:         false`.

The only difference is parenthesis around `false`. This is not a good behavior because if you miss a `||` or `&&` after `if true` and if you intend the two boolean expressions to be connected (like `if true || (false)`) , it will be hard to spot if the two expressions are long and complicated -- potential for subtle bugs.

### Expected behaviour

E492 should be reported when there is a free standing boolean expression.

### Version of Vim

9010650

### Environment

MacOS brew install.

**Comment by girishji** [comment]
Here is an example. Say, you intend the following logic:
```
 if x == 4 &&
        (y == 5 || z == 6)  
        # do something
 endif
```
Instead, you have a typo and `&&` is missing. You end up with:
```
 if x == 4
        (y == 5 || z == 6)        <-- No error
        # do something
 endif
```
This will **not** give any error. It simply ignores `(y == 5 || z == 6)`. The bug can be hard to spot depending on how complex these expressions are, but that is not the point. This should be reported as error, just as the following is an error:
```
 if x == 4
        y == 5 || z == 6         <-- Error
        # do something
 endif
```
Try this:
```
❯ vim -Nu NONE -S <(cat <<EOF
    vim9script
    :set shortmess=I
    var x = 4
    var y = 5
    var z = 6
    if x == 4
        (y == 5 || z == 6)  
        echom 'here'
    endif
EOF
)
```

**Comment by h-east** [comment]
Additional information.
There is no error with the following code either.

```vim
vim9script
if true
	[false]
	echom 'here'
endif
```

**Comment by lirenlin** [comment]
thanks for the example, I see the problem now.
It seems, there are two issues here.

one is about the the syntax, I don't think vim script has an explicit syntax for the termination of condition?
It depends on the context (the operators) or explicit line continuation.

the other thing is about the error message for expression without side-effect.
this is new in vim9. https://vim-jp.org/vimdoc-en/eval.html#E1207

:vim9cmd eval v:true --> **E1207**: Expression without an effect.
:vim9cmd eval true --> **E1207**: Expression without an effect.
:vim9cmd eval (v:true) --> no error
:vim9cmd eval [false] --> no error
:vim9cmd eval true && false --> no error

in the script, it is a little bit more complicated, it will try to interpret some expression as command
and give this error: **E492**: Not an editor command

**Comment by h-east** [comment]
That's right. The following two expressions are grammatically correct. However, there are no side effects.

```
(true)
[false]
```

Document From `:h expression-syntax`

```
2. Expression syntax					*expression-syntax*
							*E1143*
Expression syntax summary, from least to most significant:

...snip...
|expr11|  number		number constant
	"string"		string constant, backslash is special
	'string'		string constant, ' is doubled
	[expr1, ...]		|List|                                        <-- in this case
	{expr1: expr1, ...}	|Dictionary|
	#{key: expr1, ...}	legacy |Dictionary|
	&option			option value
	(expr1)			nested expression                             <-- in this case
...snip...
```

---

In the case of a string constant, `E1207` occurs, so it might be good to follow this.
```
:vim9cmd eval "str"
E1207: Expression without an effect: "str"
```

**Comment by lirenlin** [comment]
https://github.com/vim/vim/blob/57b947e3c3b8d52b914158979263855d785445d5/src/vim9cmds.c#L1955
https://github.com/vim/vim/blob/57b947e3c3b8d52b914158979263855d785445d5/src/ex_eval.c#L937

the checking logic only works for the simplest expression (kind of Terminal symbols): constant, literal, variable name, etc. 
more complicated product is not considered.
